core-processor/src/main/java/io/micronaut/expressions/parser/ast/ExpressionNode.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2017-2022 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.expressions.parser.ast;
+
+import io.micronaut.expressions.parser.compilation.ExpressionCompilationContext;
+import org.objectweb.asm.Type;
+
+/**
+ * Abstract evaluated expression AST node.
+ *
+ * @author Sergey Gavrilov
+ * @since 4.0.0
+ */
+public abstract class ExpressionNode {
```

Comment:
can this be a sealed class

Reply:
It would require all subtypes of `ExpressionNode` to be in the same package and I would prefer storing them separately in logically grouped packages. However, I've added sealed classes down the hierarchy for AST nodes such as `BinaryOperator`, `MathOperator` etc.